### PR TITLE
[14.0][IMP][FIX] account_invoice_import_invoice2data add line support and additional partner fields

### PR DESCRIPTION
**Pull request by bosd**
[a121bb6d-91b3-4233-b89a-af9dd03a6cf1.webm](https://user-images.githubusercontent.com/11499387/231365074-3e690dee-9345-4798-94f9-ad7bd1ef6ffb.webm)

add support for:
Invoice lines

- [x] `discount` 
- [x] `price_subtotal` 
- [x] `date_start` + `date_end`
- [x] `uom_dict`
    - [x] `unece_code`
    - [x] `name` 
- [x] Product dict 
    - [x] `barcode`
    - [x] `code`
- [x] tax dict 
    - [x] `amount`
    - [x]  `vat_percent`
    - [x] `price_include` True of False (preparation for future Support for ATI All tax included) 
- [x] Invoice line notes // upstream support in account_invoice_import to be implemented
- [x] Invoice line sectionheader // upstream support in account_invoice_import to be implemented

partner fields:
- [x] Chamber of Commerce Number
- [x] bic
- [x] iban
- [x] note
- [x] phone
- [x] street
- [x] street2
- [x] city
- [x] zip
- [x] country_code

Global invoice fields:
- [x] note (will be imported in the chatter)
- [x] narration  (no support yet in account_invoice_import) Will be imported in bottom of invoice page

- [x] Improve documentation

todo:
- [x] add unit tests and test invoice

Leave other keys and values in tact so they could be processed by addons downstream.

related to issue: https://github.com/OCA/edi/issues/568

Keep compatability with template builder: https://github.com/OCA/edi/pull/399
Use standard field names: https://github.com/invoice-x/invoice2data/pull/313#issuecomment-1079958934

### CODE REVIEW

**Comment by bosd** [comment]
I'd say this one is ready for review.
Some tests could be added. 
Maybe a small improvement, in case there is no partner name passed  from the invoice2data library, use the issuer field instead.

FYI, I'm currently working on adding section header and note lines support in the account invoice import module.

**Comment by bosd** [comment]
Please remove the stale label. I am still working on this. Even in conjunction with invoice2data lib.

Will add some tests/examples

**Comment by bosd** [comment]
All green!! It's ready now,
@MTantin, @bofiltd, @bealdav Can you please review?

**Review by bofiltd — APPROVED** [review]
Code review + Functional test LGTM :+1:

**Comment by bosd** [comment]
@etobella Can you review? :pray:

**Comment by bealdav** [comment]
@bosd I'm not familiar with this module, I need some time to dive in your PR. I'll try.

Thanks for this PR, code and nice doc

**Review by etobella — APPROVED** [review]
LGTM. Technical review.... You added some non existant fields, but I assume this data will not be imported unless you have that modules installed

**Comment by bosd** [comment]
@alexis-via Can you merge this?

**Comment by bosd** [comment]
@simahawk Can you merge this? :pray:

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py [hunk 1]
```diff
@@ -92,22 +122,84 @@ def invoice2data_parse_invoice(self, file_data):
 
     @api.model
     def invoice2data_to_parsed_inv(self, invoice2data_res):
+        lines = invoice2data_res.get("lines", [])
+
+        for line in lines:
+            # Manipulate line data to match with account_invoice_import
+            line["price_unit"] = float(line.get("price_unit", 0))
```
Comment (simahawk) [inline account_invoice_import_invoice2data/wizard/account_invoice_import.py:132]: Would be nice to move this to its own method for preparing lines

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py [hunk 2]
```diff
@@ -92,22 +122,84 @@ def invoice2data_parse_invoice(self, file_data):
...
+            line["sectionheader"] = line.get("sectionheader")
+            # qty 0 should be allowed to import notes, but not supported by document_import
+            line["qty"] = float(line.get("qty", 1))
+            if line["qty"] > 0:
+                uom_dict = {
+                    "unece_code": line.get("unece_code"),
+                    "name": line.get("uom"),
+                }
+                line["uom"] = uom_dict
+                if line.get("date_start"):
```
Comment (simahawk) [inline account_invoice_import_invoice2data/wizard/account_invoice_import.py]: this block seems duplicated, see line 137

**Review by simahawk — COMMENTED** [review]
Non blocking changes since this PR is here since quite some time.
However, pls consider them if you have time :wink:

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py (hunk 2 above)
Comment (bosd) [reply account_invoice_import_invoice2data/wizard/account_invoice_import.py]: Yes, For now I only removed the duplicates.
Maybe later I'll do the refactor in other pr. 
But there is more functionality I want to add first. in a different pr.

**Comment by bosd** [comment]
@simahawk I fixed the last commit message. Can you merge this?

**Comment by bosd** [comment]
ping @simahawk

**Comment by bosd** [comment]
Missed some fields in the documentation. Force pushed..

**Comment by simahawk** [comment]
Build red in 14.0. Reported on https://github.com/OCA/edi/issues/818

**Comment by bosd** [comment]
👏 All green now.. After some while using this in production. A few usability and doc improvements have been made.
Is this one a go now? 🙏

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py (hunk 1 above)
Comment (bosd) [reply account_invoice_import_invoice2data/wizard/account_invoice_import.py:132]: Done :tada:

**Comment by bosd** [comment]
Hmm, I keep fighting the runboat.
On local system, demo data is loaded fine.
Runboat still does'nt want to load it.

temporarily gonna comment out the line.. 
It will create an extra adjustment line on the imported invoice of AzureInterior. No biggy.

**Comment by Dranyel-Bosd** [comment]
[Uploading Screencast from 2023-10-09 16-31-09.webm…]()

**Comment by Dranyel-Bosd** [comment]
[Screencast from 2023-10-09 16-31-09.webm](https://github.com/OCA/edi/assets/147027997/a2b74a29-6648-47ee-be93-4132ca209579)

**Review by Dranyel-Bosd — APPROVED** [review]
LGTM :+1: 

[Screencast from 2023-10-09 16-31-09.webm](https://github.com/OCA/edi/assets/147027997/9891d46e-db00-4b62-ae4d-301fcfce6367)

File: account_invoice_import_invoice2data/readme/CONFIGURE.rst [hunk 3]
```diff
@@ -27,3 +29,76 @@ On the output, you will get first the text of the PDF, then some debug info on t
   * 'date': the date of the invoice
   * 'invoice_number'
   * 'date_due', if this information is available in the text of the PDF file.
+
+The invoice2data library is quite powerfull. It supports multiple input methods (pdftotext, ocrmypdf, tesseract ocr, google cloud vision).
```
Comment (simahawk) [inline account_invoice_import_invoice2data/readme/CONFIGURE.rst:33]:
```suggestion
The invoice2data library is quite powerful. It supports multiple input methods (pdftotext, ocrmypdf, tesseract ocr, google cloud vision).
```
This seems more a collateral note. You can maybe add a section "Dependencies"

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py [hunk 4]
```diff
@@ -90,24 +120,89 @@ def invoice2data_parse_invoice(self, file_data):
...
+
     @api.model
     def invoice2data_to_parsed_inv(self, invoice2data_res):
+        lines = invoice2data_res.get("lines", [])
+        prepared_lines = self.invoice2data_prepare_lines(lines)
+
         parsed_inv = {
             "partner": {
-                "vat": invoice2data_res.get("vat"),
+                "vat": re.sub(r"\W+", "", invoice2data_res.get("vat", "")).upper(),
```
Comment (simahawk) [inline account_invoice_import_invoice2data/wizard/account_invoice_import.py]: I'd move this out to `_get_vat` or something similar. Then you can add a unit test :wink:

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py [hunk 5]
```diff
@@ -90,24 +120,89 @@ def invoice2data_parse_invoice(self, file_data):
...
                 "website": invoice2data_res.get("partner_website"),
+                "phone": invoice2data_res.get("telephone"),
+                "mobile": invoice2data_res.get("mobile"),
+                "ref": invoice2data_res.get("partner_ref"),
                 "siren": invoice2data_res.get("siren"),
+                "coc_registration_number": re.sub(
+                    r"\D+", "", invoice2data_res.get("partner_coc", "")
+                ),
             },
+            "bic": re.sub(r"\W+", "", invoice2data_res.get("bic", "")).upper(),
```
Comment (simahawk) [inline account_invoice_import_invoice2data/wizard/account_invoice_import.py]: hmm ok, you don't need a `_get_vat` you need a `_clean_string` or similar :)

**Review by simahawk — APPROVED** [review]
LG overall. Just code review.
Some small remarks.
Pre-approving.

File: account_invoice_import_invoice2data/wizard/account_invoice_import.py (hunk 4 above)
Comment (bosd) [reply account_invoice_import_invoice2data/wizard/account_invoice_import.py]: Actually the convert to upper might no longer be needed.
It's already covered here.

https://github.com/OCA/edi/blob/14.0/base_business_document_import/models/business_document_import.py#L312-L313

File: account_invoice_import_invoice2data/readme/CONFIGURE.rst (hunk 3 above)
Comment (bosd) [reply account_invoice_import_invoice2data/readme/CONFIGURE.rst:33]: Done, added dependencies section  :+1:

**Comment by bosd** [comment]
@oca/edi-maintainers Can we get a merge?

**Pull request merged**